### PR TITLE
Add ChoiceFlag

### PR DESCRIPTION
**Pull request by Poweranimal**
A cli flag used for enums.
The flag displays all possible enum variations in the cli's help message and verifies the provided input value.

## What type of PR is this?

- feature

## What this PR does / why we need it:

This PR adds an enum flag implementation.

## Special notes for your reviewer:

I'm open for any suggestion to fulfill the job of supporting an enum flag better.

## Example
```
type LogFormat int32

const (
    LogFormatJSON LogFormat = iota
    LogFormatText
)

type LogMode in32

const (
    LogModeNoop LogMode = iota
    LogModeStdout
)

var logModeMap = map[LogMode]string{
    LogModeNoop: "noop",
    LogModeStdout: "stdout",
}

func (m LogMode) String() string {
    return logModeMap[m]
}

var (
    destFormat LogFormat
    destLevel string
)

var LogFormatFlag = &cli.ChoiceFlag{
    Name:         "log-format",
    Value:          LogFormatJSON,
    Choice:        cli.NewChoice(map[string]interface{}{
        "json": LogFormatJSON,
        "text":  LogFormatText,
    },
    Destination: &destFormat,
    EnvVars:      []string{"LOG_FORMAT"},
    Usage:         "The output formatting of the logger.",
    Required:     false,
}

var LogModeFlag = &cli.ChoiceFlag{
    Name:         "log-mode",
    Value:          LogModeStdout,
    Choice:        cli.NewStringerChoice(LogModeNoop, LogModeStdout),
    EnvVars:      []string{"LOG_MODE"},
    Usage:         "The log mode of the logger.",
}

var LogLevelFlag = &cli.ChoiceFlag{
    Name:         "log-level",
    Value:          "info",
    Choice:        cli.NewStringChoice("debug", "info", "warn", "error"),
    Destination: &destLevel,
    EnvVars:      []string{"LOG_LEVEL"},
    Usage:         "The log level of the logger.",
    Required:     false,
}
```

## Release Notes

```release-note

Added `cli.ChoiceFlag` for enum support.
```

## TODOS

- [x] Typed destination
- [ ] Add tests
- [ ] Add README

### CODE REVIEW

File: flag_enum.go
Comment (dearchap) [inline flag_enum.go:12]: What happens if h is nil ? i.e user requests an enum with invalid name ?

File: flag_enum.go
Comment (dearchap) [inline flag_enum.go:70]: Can you add a Destination to *Enum as well ?

**Review by dearchap — COMMENTED** [review]
Great good.

**Comment by dearchap** [comment]
I just realised. This is more generic than enums. You could implement an option of choices for strings, ints, etc for this. Maybe rename this to ChoiceFlag ?

File: flag_enum.go
Comment (Poweranimal) [reply flag_enum.go]: I haven't payed attention to it.
I would suggest returning nil, if the name is invalid.
Doing so it would correspond to the existing api of other flags.

**Comment by Poweranimal** [comment]
> I just realised. This is more generic than enums. You could implement an option of choices for strings, ints, etc for this. Maybe rename this to ChoiceFlag ?

Yeah, that sounds better.
I renamed 'enum' to 'choice'

**Comment by Poweranimal** [comment]
@dearchap Regarding adding a Destination.
This seems a bit more complicated.
As of now, this package uses for destination the `flag` package of the golang stdlib.
It doesn't play well with the `ChoiceHolder`.
Let me see, if I figure out how one can support this...

**Comment by Poweranimal** [comment]
I added destination to the ChoiceFlag and improved its API.

File: flag_choice.go
Comment (dearchap) [inline flag_choice.go:174]: remove debug statement

**Review by dearchap — COMMENTED** [review]
You should also write some unit tests for this for completion.

**Comment by Poweranimal** [comment]
I did some changes to the ChoiceFlag;

- Replaced `ChoiceHolder` with `interface{}`
- Renamed `ChoiceDecoder` to `Choice`
- Added `ToString` to `Choice`
- Added helper `Choice` constructors: `NewChoiceDecoder` and `NewStringChoiceDecoder`

**Review by marcofranssen — CHANGES_REQUESTED** [review]
We might also want to update the README.md with usage example and documenting the availabiltiy of this param type.

**Comment by marcofranssen** [comment]
What is the state of this PR? Seems only blocked on some missing code coverage for the new feature. @Poweranimal is that something you could finalize?

**Comment by Poweranimal** [comment]
Hi @marcofranssen 
I got quite demotivated to continue due to the lack of engagement by the maintainers.

Thus, I got the impression that this project is EOL and doesn't welcome pr's anymore.

**Comment by abdennour** [comment]
@Poweranimal  , you did great job. I hope better response from maintainers

**Comment by meatballhat** [comment]
@Poweranimal Hello and sorry for the long delay 😭 Please come back! This whole global pandemic thing and various maintainers' life adventures have seriously impacted responsiveness, but we're trying to get stuff flowing again now.

**Comment by meatballhat** [comment]
Connected to #602

**Comment by jalavosus** [comment]
@Poweranimal if you could rebase this onto the current `main` branch, I'd love to review this and get it merged.

**Comment by Poweranimal** [comment]
Hi @jalavosus ,
I’ve quit using this repo for quite l long time now.
Please feel free to to reuse my branch as you like.

**Comment by dearchap** [comment]
@meatballhat I think this is a good feature but there are many ways to go about it. Either the ChoiceFlag as given here or a Choice interface can be added to each flag(IntFlag, StringFlag etc) in the generated code. I am inkling towards the latter approach. I can work on it if you think its the right direction.

**Comment by meatballhat** [comment]
@dearchap I would love to see what you have in mind for a Choice interface!

**Comment by dearchap** [comment]
This approach has been considered but dropped due to favoring a different mechanism using a validation API. That would allow ALL flags to use validation code instead of creating a new Flag

**Pull request closed without merging**
